### PR TITLE
[web] Initial rich measurement implementation

### PR DESCRIPTION
**Pull request by mdebbar**
## Description

This is the first in a sequence of PRs to implement rich text measurement. It performs measurement of rich text that doesn't contain ellipsis nor placeholders. The place for handling ellipsis and placeholders is marked with TODOs inside the code, and will be added in subsequent PRs.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+      // Only go to the next span if we've reached the end of this span.
+      if (currentLine.end.index >= span.end && spanIndex < spanCount - 1) {
+        span = paragraph.spans[++spanIndex];
+      }
+    }
+  }
+}
+
+class LineBuilder {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:182]: Please add doc comment. Specifically explain usage. Calling LineBuilder.first() and then calling nextLine() as you iterate through spans.

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    return LineBuilder._(
+      paragraph,
+      spanometer,
+      start: end,
+      lineNumber: lineNumber + 1,
+    );
+  }
+}
+
+class Spanometer {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:273]: Doc comment

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+  final List<EngineLineMetrics> lines = <EngineLineMetrics>[];
+
+  // *** Convenient shortcuts used during layout *** //
+
+  int? get maxLines => paragraph.paragraphStyle._maxLines;
+  bool get unlimitedLines => maxLines == null;
+  bool get hasEllipsis => paragraph.paragraphStyle._ellipsis != null;
+
+  void perform(ui.ParagraphConstraints constraints) {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:46]: suggestion: performLayout. needs doc comment

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:208]: Done

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:320]: Done

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 3 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart]: Done

**Review by ferhatb — APPROVED** [review]

**Pull request merged**